wgpu/tests/wasm.rs

```diff
@@ -0,0 +1,210 @@
+        width: size.width,
+        height: size.height,
+        present_mode: wgpu::PresentMode::Fifo,
+        alpha_mode: surface.get_supported_alpha_modes(&adapter)[0],
+    };
+
+    surface.configure(&device, &config);
+
+    event_loop.spawn(move |event, _, control_flow| {
+        *control_flow = ControlFlow::Wait;
+        if let Event::RedrawRequested(_) = event {
+            let frame = surface
+                .get_current_texture()
+                .expect("Failed to acquire next swap chain texture");
+            let view = frame
+                .texture
+                .create_view(&wgpu::TextureViewDescriptor::default());
+            let mut encoder =
+                device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+            {
+                let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+                    label: None,
+                    color_attachments: &[Some(wgpu::RenderPassColorAttachment {
+                        view: &view,
+                        resolve_target: None,
+                        ops: wgpu::Operations {
+                            load: wgpu::LoadOp::Clear(wgpu::Color::GREEN),
+                            store: true,
+                        },
+                    })],
+                    depth_stencil_attachment: None,
+                });
+                rpass.set_pipeline(&render_pipeline);
+                rpass.draw(0..3, 0..1);
+            }
+
+            queue.submit(Some(encoder.finish()));
+            frame.present();
+
+            assert_rendering_result(&window);
+
+            *control_flow = ControlFlow::Exit;
+        }
+    });
+}
+
+fn read_pixel(canvas: HtmlCanvasElement, x: i32, y: i32) -> [u8; 4] {
+    let mut result = [0_u8; 4];
+    let context = canvas
+        .get_context("webgl2")
+        .unwrap()
+        .unwrap()
+        .dyn_into::<web_sys::WebGl2RenderingContext>()
+        .unwrap();
+
+    context
+        .read_pixels_with_u8_array_and_dst_offset(
+            x,
+            y,
+            1,
+            1,
+            web_sys::WebGl2RenderingContext::RGBA,
+            web_sys::WebGl2RenderingContext::UNSIGNED_BYTE,
+            &mut result,
+            0,
+        )
+        .unwrap();
+    result
+}
+
+async fn init() -> (Window, EventLoop<()>, Surface, Adapter, Device, Queue) {
+    let event_loop = EventLoop::new();
+    let window = winit::window::Window::new(&event_loop).unwrap();
+
+    std::panic::set_hook(Box::new(console_error_panic_hook::hook));
+    console_log::init().expect("could not initialize logger");
+    // On wasm, append the canvas to the document body
+    web_sys::window()
+        .and_then(|win| win.document())
+        .and_then(|doc| doc.body())
+        .and_then(|body| {
+            body.append_child(&web_sys::Element::from(window.canvas()))
+                .ok()
+        })
+        .expect("couldn't append canvas to document body");
```

Comment:
The end goal here is to get this logic into initialize_test in tests/common. This way all our existing tests can run on wasm without modification of the test itself.